cpp/cert/src/rules/CTR55-CPP/DoNotUseAnAdditiveOperatorOnAnIterator.ql

```diff
@@ -14,24 +14,65 @@
 import cpp
 import codingstandards.cpp.cert
 import codingstandards.cpp.Iterators
+import semmle.code.cpp.controlflow.Dominance
 
-from ContainerIteratorAccess it
+/**
+ * something like:
+ * `end = begin() + size()`
+ */
+Expr calculatedEndCheck(AdditiveOperatorFunctionCall calc) {
+  exists(
+    ContainerAccessWithoutRangeCheck::ContainerSizeCall size,
+    ContainerAccessWithoutRangeCheck::ContainerBeginCall begin
+  |
+    calc.getTarget().hasName("operator+") and
+    DataFlow::localFlow(DataFlow::exprNode(size), DataFlow::exprNode(calc.getAChild*())) and
+    DataFlow::localFlow(DataFlow::exprNode(begin), DataFlow::exprNode(calc.getAChild*())) and
+    //make sure its the same container providing its size as giving the begin
+    globalValueNumber(begin.getQualifier()) = globalValueNumber(size.getQualifier()) and
+    result = begin.getQualifier()
+  )
+}
+
+Expr validEndCheck(FunctionCall end) {
+  end instanceof ContainerAccessWithoutRangeCheck::ContainerEndCall and
+  result = end.getQualifier()
+  or
+  result = calculatedEndCheck(end)
+}
+
+/**
+ * some guard exists like: `iterator != end`
+ * where a relevant`.end()` call flowed into end
+ */
+predicate validEndBoundCheck(ContainerIteratorAccess it, IteratorSource source) {
+  exists(
+    FunctionCall end, BasicBlock b, GuardCondition l, ContainerIteratorAccess otherAccess,
+    Expr qualifierToCheck
+  |
+    //sufficient end guard
+    qualifierToCheck = validEndCheck(end) and
+    //guard controls the access
+    l.controls(b, _) and
+    b.contains(it) and
+    //guard is comprised of end check and an iterator access
+    DataFlow::localFlow(DataFlow::exprNode(end), DataFlow::exprNode(l.getChild(_))) and
+    l.getChild(_) = otherAccess and
+    //make sure its the same iterator being checked in the guard as accessed
+    otherAccess.getOwningContainer() = it.getOwningContainer() and
```

Comment:
This seems to not account for the case where two different iterators are created from the same container.

I don't think `globalValueNumber` will be useful because the iterator access might have a different value due to crements.

For now it is probably sufficient 🤔.

Reply:
globalValueNumber here isnt checking the iterator accesses themselves, its checking the qualifier of the container function accesses (ie crements not relevant)

so for this:
```
   {
   auto end = v.end(); //line 1
    for (auto i = v.begin(); i != end; ++i) { // COMPLIANT //line 2
    }
  }
```
it will be checking is the v on line 1 is the same as the v on line 2

it cuts out cases like this, where the vector end check is not for the corresponding iterator access
```  
void test(std::vector<int> &v, std::vector<int> &v2) {
{
   auto end = v2.end(); //line 1
    for (auto i = v.begin(); i != end; ++i) { // NON_COMPLIANT
    }
  }
}

```

Reply:
My comment might have confusing, but I was not referring to the current use of `globalValueNumber`, but the potential use to determine if `otherAccess` and `it` are the same iterator. Our current check isn't sufficient in the case of multiple iterators.

Reply:
hm, 
this case seems to me to be more corner case, or at least I am having trouble thinking of an example where this seems common/makes sense
feel free to suggest one if you have one in mind! otherwise, we can leave as a future improvement if we see cases where this misses out

Reply:
It is definitely a corner case, so I don't think we need to solve now.